core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/deletevolumes/DeleteVolumesActions.java

```diff
@@ -0,0 +1,221 @@
+                        .map(s -> cloudResourceConverter.convert(s)).collect(toList());
+                    String cloudPlatform = stack.getCloudPlatform();
+                    DeleteVolumesRequest deleteVolumesRequest = new DeleteVolumesRequest(cloudResourcesToBeDeleted, stackDeleteVolumesRequest, cloudPlatform);
+                    sendEvent(context, DELETE_VOLUMES_EVENT.selector(), deleteVolumesRequest);
+                }
+            }
+        };
+    }
+
+    @Bean(name = "DELETE_VOLUMES_FINISHED_STATE")
+    public Action<?, ?> deleteVolumesFinishedAction() {
+        return new AbstractClusterAction<>(DeleteVolumesRequest.class) {
+
+            @Override
+            protected void doExecute(ClusterViewContext context, DeleteVolumesRequest payload, Map<Object, Object> variables) {
+                StackDeleteVolumesRequest stackDeleteVolumesRequest = payload.getStackDeleteVolumesRequest();
+                StackDto stack = (StackDto) variables.get("STACK_DTO");
+                boolean detachedVolumes = false;
+                String cloudPlatform = payload.getCloudPlatform();
+                CloudPlatformVariant cloudPlatformVariant = new CloudPlatformVariant(Platform.platform(cloudPlatform), Variant.variant(cloudPlatform));
+                CloudConnector cloudConnector = cloudPlatformConnectors.get(cloudPlatformVariant);
+                CloudCredential cloudCredential = stackUtil.getCloudCredential(stack.getEnvironmentCrn());
+                Location location = location(region(stack.getRegion()), availabilityZone(stack.getAvailabilityZone()));
+                CloudContext cloudContext = CloudContext.Builder.builder()
+                        .withId(stack.getId())
+                        .withName(stack.getName())
+                        .withCrn(stack.getResourceCrn())
+                        .withPlatform(stack.getCloudPlatform())
+                        .withVariant(stack.getPlatformVariant())
+                        .withLocation(location)
+                        .withWorkspaceId(stack.getWorkspace().getId())
+                        .withAccountId(Crn.safeFromString(stack.getResourceCrn()).getAccountId())
+                        .build();
+                AuthenticatedContext ac = getAuthenticatedContext(cloudCredential, cloudContext, cloudConnector);
+                List<CloudResource> cloudResourcesToBeDeleted = payload.getResourcesToBeDeleted();
+                try {
+                    deleteVolumesService.detachResources(cloudResourcesToBeDeleted, cloudPlatform, ac);
+                    deleteVolumesService.updateCbdbResources(stack, payload);
+                    detachedVolumes = true;
+                    DeleteVolumesFinalizedEvent event = new DeleteVolumesFinalizedEvent(payload.getStackDeleteVolumesRequest().getStackId());
+                    sendEvent(context, event);
+                } catch (Exception ex) {
```

Comment:
Why are we sending final event before deleting the volumes? This will cause transition to final state while operation is still in progress. I think, It can cause issues. Anyways, We are handling failures for delete gracefully so we should final event after deleting the volumes

Reply:
changed